network/crl.go

```diff
@@ -0,0 +1,370 @@
+		httpRequest.Header.Add("host", crlURL.Host)
+		httpResponse, err := c.httpClient.Do(httpRequest)
+		if err != nil {
+			return nil, err
+		}
+		defer httpResponse.Body.Close()
+		content, err := ioutil.ReadAll(httpResponse.Body)
+		if err != nil {
+			return nil, err
+		}
+
+		if httpResponse.StatusCode != 200 {
+			log.WithField("status", httpResponse.Status).Warnln("Server returned non-OK status")
+			return nil, ErrHTTPServerReturnedError
+		}
+
+		return content, nil
+	} else if strings.HasPrefix(uri, "file://") {
+		content, err := ioutil.ReadFile(uri[7:])
+		if err != nil {
+			return nil, err
+		}
+
+		return content, nil
+	}
+
+	log.WithField("uri", uri).Warnln("Cannot fetch CRL")
+	return nil, ErrFetchCRLUnsupportedProtocol
+}
+
+// CRLCacheItem is combination of fetched+parsed+verified CRL with fetch time
+type CRLCacheItem struct {
+	Fetched time.Time // When this CRL was fetched and cached
+	CRL     pkix.CertificateList
+}
+
+// CRLCache is used to store fetched CRLs to avoid downloading the same URI more than once,
+// stores parsed and verified CRLs
+type CRLCache interface {
+	Get(key string) (*CRLCacheItem, error)
+	Put(key string, value *CRLCacheItem) error
+	Remove(key string) error
+}
+
+// LRUCRLCache is an implementation of CRLCache that uses LRU cache inside
+type LRUCRLCache struct {
+	cache lru.Cache
+	mutex sync.RWMutex
+}
+
+// NewLRUCRLCache creates new LRUCRLCache, able to store at most maxEntries values
+func NewLRUCRLCache(maxEntries int) *LRUCRLCache {
+	return &LRUCRLCache{cache: lru.Cache{MaxEntries: maxEntries}}
+}
+
+// Get tries to get CRL from cache, returns error if failed
+func (c *LRUCRLCache) Get(key string) (*CRLCacheItem, error) {
+	log.Debugf("CRL: LRU cache: loading '%s'", key)
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	value, ok := c.cache.Get(key)
+	if ok {
+		value, _ := value.(*CRLCacheItem)
+		log.Debugf("CRL: LRU cache: '%s' found", key)
+		return value, nil
+	}
+
+	return nil, ErrCacheKeyNotFound
+}
+
+// Put stores CRL in cache
+func (c *LRUCRLCache) Put(key string, value *CRLCacheItem) error {
+	c.mutex.Lock()
+	c.cache.Add(key, value)
+	c.mutex.Unlock()
+	log.Debugf("CRL: LRU cache: inserted '%s'", key)
+	return nil
+}
+
+// Remove removes item from cache
+func (c *LRUCRLCache) Remove(key string) error {
+	c.mutex.Lock()
+	c.cache.Remove(key)
+	c.mutex.Unlock()
+	log.Debugf("CRL: LRU cache: removed '%s'", key)
+	return nil
+}
+
+// DefaultCRLVerifier is a default implementation of CRLVerifier
+type DefaultCRLVerifier struct {
+	Config CRLConfig
+	Client CRLClient
+	Cache  CRLCache
+}
+
+// Tries to find cached CRL, fetches using v.Client if not found, checks the signature of CRL using issuerCert
+func (v DefaultCRLVerifier) getCachedOrFetch(uri string, issuerCert *x509.Certificate) (*pkix.CertificateList, error) {
+	// Try v.Cache first, but only if caching is enabled (cache time > 0)
+	if v.Config.cacheTime > 0 {
```

Comment:
Currently, cache size is just passed to LRU cache constructor, and `0` there means "unlimited capacity" rather than "no capacity". I guess, we should disallow `0` and tell users "disable caching by setting time to `0`"? Or, handle zero cache size as "cache disabler", but then there's s question about what value should cache time have in this case.

Reply:
>  I guess, we should disallow 0 and tell users "disable caching by setting time to 0"? Or, handle zero cache size as "cache disabler", but then there's s question about what value should cache time have in this case.

I guess both work and are predictable enough. If either cache size or cache duration is zero then caching is disabled. If it is disabled due to one value being zero then the other value is more or less irrelevant, as long as it can be parsed, I guess.

Reply:
So, created
```go
func (c *CRLConfig) isCachingEnabled() bool {
	return c.cacheTime > 0 && c.cacheSize > 0
}
```
and using it here instead `if v.Config.cacheTime > 0`